primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -181,10 +188,387 @@ where
+        Ok(NMT {
+            inner,
+            namespace_ranges,
+        })
+    }
+
+    fn height(&self) -> usize {
+        self.inner.height()
+    }
+
+    fn capacity(&self) -> num_bigint::BigUint {
+        self.inner.capacity()
+    }
+
+    fn num_leaves(&self) -> u64 {
+        self.inner.num_leaves()
+    }
+
+    fn commitment(&self) -> Self::Commitment {
+        self.inner.commitment()
+    }
+
+    fn lookup(
+        &self,
+        pos: impl Borrow<Self::Index>,
+    ) -> super::LookupResult<Self::Element, Self::MembershipProof, ()> {
+        self.inner.lookup(pos)
+    }
+
+    fn verify(
+        root: impl Borrow<Self::NodeValue>,
+        pos: impl Borrow<Self::Index>,
+        proof: impl Borrow<Self::MembershipProof>,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        <InnerTree<E, H, T, N, Arity> as MerkleTreeScheme>::verify(root, pos, proof)
+    }
+}
+
+impl<E, H, Arity, N, T> AppendableMerkleTreeScheme for NMT<E, H, Arity, N, T>
+where
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    fn extend(
+        &mut self,
+        elems: impl IntoIterator<Item = impl core::borrow::Borrow<Self::Element>>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO: update namespace metadata
+        self.inner.extend(elems)
+    }
+
+    fn push(
+        &mut self,
+        elem: impl core::borrow::Borrow<Self::Element>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO: update namespace metadata
+        self.inner.push(elem)
+    }
+}
+
+#[derive(Clone, Debug)]
+enum NamespaceProofType {
+    Presence,
+    Absence,
+}
+
+#[derive(Clone, Debug)]
+/// Namespace Proof
+pub struct NamespaceProof<E, T, Arity, N, H>
+where
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    proof_type: NamespaceProofType,
+    proofs: Vec<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    left_boundary_proof: Option<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    right_boundary_proof: Option<MerkleProof<E, u64, NamespacedHash<T, N>, Arity>>,
+    indices: Vec<u64>,
+    phantom: PhantomData<H>,
+}
+impl<E, T, Arity, N, H> NamespaceProof<E, T, Arity, N, H>
+where
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    /// Return the set of leaves associated with this Namespace proof as an
+    /// iterator
+    pub fn get_namespace_leaves(&self) -> impl Iterator<Item = &E> {
+        let num_leaves = match self.proof_type {
+            NamespaceProofType::Presence => self.proofs.len(),
+            NamespaceProofType::Absence => 0,
+        };
+        self.proofs
+            .iter()
+            .map(|proof| proof.elem().unwrap())
```

Comment:
Assuming the namespace proof is valid, this unwrap should be safe (the proof checks that each elem/leaf is accounted for). We could verify this here as well, but I think that verifying the namespace proof before fetching the elements is a reasonable precondition.

Reply:
we could consider removing `get_namespace_leaves` (or making it private) and having `verify` return the leaves on success. So you have to verify in order to get the leaves

Reply:
good idea

Reply:
Looks this would require refactoring `VerificationResult` from a `Result<(),()>` to a `Result<T, ()>`, which is think is a reasonable improvement (cc. @ggutoski) but probably something to tackle in another PR

Reply:
Yes almost always a custom result type alias should be generic over the `Ok` type. But yeah we can save this for future work